cosmos_rl/rollout/vllm_rollout/vllm_rollout_worker_async.py

```diff
@@ -0,0 +1,1469 @@
+                        validation_step=current_step,
+                        prompt_idxs=[],
+                        payloads=payloads,
+                        is_end=True,
+                    )
+                    self.api_client.post_validation_report(response)
+                payloads, is_validation, current_step, empty = (
+                    self.reward_dispatcher.dequeue_rewards_cal()
+                )
+
+    @torch.no_grad()
+    async def do_generate(self):
+        logger.debug(f"[Rollout] generate start for rank {self.global_rank}")
+
+        # we enqueue tasks by batch in the main loop, so we need to collect all task results here.
+        await self.scheduler.wait_all_tasks_completed()
+        rollout_results = self.scheduler.get_all()
+
+        assert (
+            len(rollout_results) == self.batch_size
+        ), f"Error: VLLM returned {len(rollout_results)} for {self.batch_size}"
+
+        # we need filter the result with valid completions or valid completed_conversations
+        valid_results: list[CompletedRollout] = []
+        if self.rollout.rollout_config.multi_turn_config.enable:
+            valid_results = filter_valid_multi_turn_rollout_results(rollout_results)
+        else:
+            valid_results = filter_valid_single_turn_rollout_results(
+                rollout_results, self.eos_token
+            )
+
+        logger.debug(f"[Rollout] generate end for rank {self.global_rank}")
+
+        should_report = (
+            self.parallel_dims.tp_coord[0] == 0
+            and (self.parallel_dims.pp_coord[0] == self.parallel_dims.pp_coord[1] - 1)
+            and len(valid_results) > 0
+        )
+
+        if should_report:
+            # only the first tp rank in the rollout replica will post the completion to the controller.
+            valid_payloads: List[RLPayload] = []
+
+            for cr in valid_results:
+                # update payload
+                cr.payload.completions = cr.result.completions
+                if self.rollout.rollout_config.multi_turn_config.enable:
+                    cr.payload.completed_conversations = (
+                        cr.result.completed_conversations
+                    )
+                valid_payloads.append(cr.payload)
+
+            self.reward_dispatcher.enqueue_rewards_cal(
+                valid_payloads,
+                False,
+                self.current_weight_version,
+            )
+
+            if (
+                self.state.prompt_fetch_end()
+                and self.scheduler.is_all_tasks_completed()
+            ):
+                self.state.set_prompt_consume_end()
+
+    async def main_loop_async(self):
+        """Main loop with async scheduler integration (coroutine version)."""
+
+        await self.lazy_initialize_rollout_engine(load_format="auto")
+
+        while not self.shutdown_signal.is_set():
+            # Sync consume command
+            await self.consume_command(cmd_pred=None)
```

Comment:
Here during we handle the command including weight sync, the rollout generation may still be in processing? In this case, well the weight changes during same rollout generation?

Reply:
I think it is okay to make weight sync even during the same rollout inference. But maybe we can add an optional to choice to force the weight sync happens only when there is no unfinished pending rollouts?